File: JavaScript/1674-Minimum Moves to Make Array Complementary.js
/**
 * @param {number[]} nums
 * @param {number} limit
 * @return {number}
 */
var minMoves = function(nums, limit) {
    let N = nums.length;
    let diffMap = new Map();
    
    for (let i=0; i<N/2; i++){
        let a = nums[i];
        let b = nums[N-1-i];
        diffMap.set(a+b, (diffMap.get(a+b)||0) - 1 );
        diffMap.set(a+b+1, (diffMap.get(a+b+1)||0) + 1 );
        let min = Math.min(a,b), max = Math.max(a,b);
        diffMap.set(min+1, (diffMap.get(min+1)||0) - 1 );
        diffMap.set(max+limit+1, (diffMap.get(max+limit+1)||0) + 1);
    }
    
    let res = N;
    let cur = N;
    for (let i=2; i<= 2*limit+1; i++){
        cur += diffMap.get(i) || 0;
        res = Math.min(res, cur);
    }
    return res;
};

File: C++/1674-Minimum Moves to Make Array Complementary.cpp
class Solution {
public:
    int minMoves(vector<int>& nums, int limit) {
        int N = nums.size();
        vector<int> dp(2*limit+1, 0);

        for (int i=0; i<N/2; i++) {
            int mn = min(nums[i], nums[N-1-i]);
            int mx = max(nums[i], nums[N-1-i]);

            dp[2] += 2;
            dp[mn+1] -= 1;
            dp[mn+mx] -= 1;
            if (mn + mx + 1 <= 2*limit) {
                dp[mn+mx+1] += 1;
            }
            if (mx + limit + 1 <= 2*limit) {
                dp[mx+limit+1] += 1;
            }            
        }
        
        int running = 0, res = INT_MAX;
        for (int num=2; num<=2*limit; num++) {
            running += dp[num];
            res = min(res, running);
        }

        return res;
    }
};